endpoints/endpoints.py

```diff
@@ -0,0 +1,1426 @@
+    Determine the log level to return based on the provided result's return code.
+
+    Args:
+        result (Fabric/Invoke result): From a Fabric/Invoke run command
+
+    Globals:
+        None
+
+    Returns:
+        str: The appropriate log level to use to describe the result
+    """
+    if result.exited != 0:
+        return "error"
+    else:
+        return "info"
+
+def remote_connection(host, user, validate = False):
+    """
+    Create a Fabric connection and open it
+
+    Args:
+       host (str): The IP address or hostname to connect to using Fabric
+       user (str): The username to connect as
+       validate (bool): Is the function being called from validation mode (which means that logging cannot be used)
+
+    Globals:
+        None
+
+    Returns:
+       an open Fabric Connection
+    """
+    connection = Connection(host = host, user = user)
+    attempts = 5
+    attempt = 0
+    while attempt < attempts:
+        try:
+            attempt += 1
+            connection.open()
+            if attempt > 1:
+                msg = "Connected to remote '%s' as user '%s' after %d attempts" % (host, user, attempt)
+                if validate:
+                    validate_comment(msg)
+                else:
+                    log.info(msg)
+            break
+        except (ssh_exception.AuthenticationException, ssh_exception.NoValidConnectionsError) as e:
+            msg = "Failed to connect to remote '%s' as user '%s' on attempt %d due to '%s'" % (host, user, attempt, str(e))
+            if validate:
+                validate_comment(msg)
+            else:
+                log.warning(msg)
+
+            if attempt == attempts:
+                msg = "Failed to connect to remote '%s' as user '%s' and maximum number of attempts (%d) has been exceeded.  Reraising exception '%s'" % (host, user, attempts, str(e))
+                if validate:
+                    validate_error(msg)
+                else:
+                    log.error(msg)
+                raise e
+            else:
+                time.sleep(1)
+    return connection
+
+def validate_log(msg):
+    """
+    Log a validation message
+
+    Args:
+        msg (str): The message to log
+
+    Globals:
+        None
+
+    Returns:
+        None
+    """
+    return print(msg)
+
+def validate_comment(msg):
+    """
+    Log a validation comment message
+
+    Args:
+        msg (str): The message to log
+
+    Globals:
+        None
+
+    Returns:
+        None
+    """
+    return print("#" + msg)
+
+def validate_error(msg):
+    """
+    Log a validation error message
+
+    Args:
+        msg (str): The message to log
+
+    Globals:
+        None
+
+    Returns:
+        None
+    """
+    return print("ERROR: " + msg)
+
+def validate_debug(msg):
+    """
+    Log a vlidation debug message
+
+    Args:
+        msg (str): The message to log
+
+    Globals:
+        None
+
+    Returns:
+        None
+    """
+    return print("# DEBUG: " + msg)
```

Comment:
refresh me on why we can't merge these validate_xxx into one and add an arg... validate_log(msg, level)

Reply:
FWIW, the logging library uses `<obj>.info(...)`, `<obj>.error(...)`, `<obj>.debug(...)` instead of `<obj>.log(..., level)` for some reason....